graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,268 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
   return std::make_pair(sorted_array, sorted_idx);
 }
 
+template <typename indptr_t, typename indices_t>
+struct AlignmentFunc {
+  static_assert(GPU_CACHE_LINE_SIZE % sizeof(indices_t) == 0);
+  const indptr_t* in_degree;
+  const int64_t* perm;
+  int64_t num_rows;
+  __host__ __device__ auto operator()(int64_t row) {
+    constexpr int num_elements = GPU_CACHE_LINE_SIZE / sizeof(indices_t);
+    return thrust::make_tuple(
+        in_degree[row],
+        (indptr_t)(in_degree[perm ? perm[row % num_rows] : row] + num_elements - 1));
+  }
+};
+
+template <typename indptr_t, typename indices_t>
+__global__ void _CSRRowWiseOneHopExtractorAlignedKernel(
+    const indptr_t hop_size, const int64_t num_rows,
+    const indptr_t* const indptr, const indptr_t* const subindptr,
+    const indptr_t* const subindptr_aligned, const indices_t* const indices,
+    indices_t* const hop, const int64_t* const perm) {
+  indptr_t tx = static_cast<indptr_t>(blockIdx.x) * blockDim.x + threadIdx.x;
+  const int stride_x = gridDim.x * blockDim.x;
+
+  while (tx < hop_size) {
+    const auto rpos_ = cuda::UpperBound(subindptr_aligned, num_rows, tx) - 1;
+    const auto rpos = perm ? perm[rpos_] : rpos_;
+    const auto out_row = subindptr[rpos];
+    const auto d = subindptr[rpos + 1] - out_row;
+    const int offset =
+        ((size_t)(indices + indptr[rpos] - subindptr_aligned[rpos_]) %
+         GPU_CACHE_LINE_SIZE) /
+        sizeof(indices_t);
+    const auto rofs = tx - subindptr_aligned[rpos_] - offset;
+    if (rofs >= 0 && rofs < d) {
+      const auto in_idx = indptr[rpos] + rofs;
+      assert((size_t)(indices + in_idx - tx) % GPU_CACHE_LINE_SIZE == 0);
+      const auto u = indices[in_idx];
+      hop[out_row + rofs] = u;
+    }
+    tx += stride_x;
+  }
+}
+
+template <typename indptr_t, typename index_t>
+struct DegreeFunc {
```

Comment:
Please add documentations for these thrust functors, including the purposes of them and the states that are modified by them.

Reply:
I did, can you check?